src/psyclone/psyir/backend/fortran.py

```diff
@@ -1087,15 +1087,34 @@ def literal_node(self, node):
         :param node: a Literal PSyIR node.
         :type node: :py:class:`psyclone.psyir.nodes.Literal`
 
-        :returns: the Fortran code as a string.
+        :returns: the Fortran code for the literal.
         :rtype: str
 
         '''
         if node.datatype.intrinsic == ScalarType.Intrinsic.BOOLEAN:
             # Booleans need to be converted to Fortran format
             result = '.' + node.value + '.'
         elif node.datatype.intrinsic == ScalarType.Intrinsic.CHARACTER:
-            result = "'{0}'".format(node.value)
+            # Need to take care with which quotation symbol to use since a
+            # character string may include quotation marks, e.g. a format
+            # specifier: "('hello',3A)". The outermost quotation marks are
+            # not stored so we have to decide whether to use ' or ".
+            single_idx = node.value.find("'")
+            if single_idx == -1:
+                # No single quotes in the string so use those
+                quote_symbol = "'"
+            else:
+                dbl_idx = node.value.find('"')
+                if dbl_idx == -1:
+                    # No double quotes in the string so use those
+                    quote_symbol = '"'
+                else:
```

Comment:
This bit of code is currently needed because we permit a character string to contain both "'" and/or '"' characters. It is not to do with supporting Fortran's method of escaping single quotes. i.e. when I say "double quotes" in the comment I mean any instance of '"' in the string. Or are you saying that the frontend should refuse any strings that themselves contain any type of quotation mark?

Reply:
Forgive me if I am missing something, but what valid Fortran character string contains both "'" and '"'? My belief is that to contain both then at least one of these must be escaped, hence my suggestion. My argument goes as follows ... the string must be within either '...' or "...". Therefore if it contains both "'" and '"' then one of these will become the termination of the string.

Reply:
My main use case for this is Fortran format strings, e.g.:

    write(*, "('Hello ', (A))") "Andy"

Does that make it clearer or am I missing something?

Reply:
Hopefully you are, otherwise I am :-). My belief is that your example format string above only has single quotes in it. The code we are discussing will see `('Hello ', (A))` and will rightly decide to output with double quotes when writing the Fortran.

Reply:
Ahhh, light dawns. I seem to be a bit slow today. I think I've got it now...